### PR TITLE
Documentation issue1529 issue1530

### PR DESCRIPTION
**Pull request by JeremyBois**
Add documentation to Troubleshooting section to explain how to solve #1529 and #1530.

@moodmosaic @rwols

### CODE REVIEW

File: docs/src/troubleshooting.md [hunk 1]
```diff
@@ -42,15 +30,78 @@ Adjusting `PATH` can differ based on the operating system and the default shell
...
+export PATH=$PATH:~/.cabal/bin
+```
+
+The complete procedure of updating the `PATH` used by Sublime Text depends on your platform and is explained [here](troubleshooting.md#updating-the-path-used-by-lsp-servers).
+
+
+### 4. Popup error `Language server <your_server_language_name> has crashed`
+Assuming that the server is actually installed, and that you can start it from your shell, this issue is likely due to Sublime Text's internal environment not picking up the same `PATH` environment variable as you've configured in your shell.
+
+Server languages depend on other binaries that should also be in your `PATH` in addtion to the server binary itself.
```
Comment (rwols) [inline docs/src/troubleshooting.md:98]: Server languages --> Language servers

Comment (rwols) [inline docs/src/troubleshooting.md:98]: depend --> may depend

File: docs/src/troubleshooting.md [hunk 2]
```diff
@@ -42,15 +30,78 @@ Adjusting `PATH` can differ based on the operating system and the default shell
...
+If, for example, you want to expose a `Node` binary to ST and you have it installed through a version manager like `nvm`, you need to insert its [initialization script](https://github.com/nvm-sh/nvm#install--update-script) in the location specified in [this table](troubleshooting.md#updating-the-path-used-by-lsp-servers)
+
+The complete procedure of updating the `PATH` used by Sublime Text depends on your platform and is explained [here](troubleshooting.md#updating-the-path-used-by-lsp-servers).
+
+### 3. Error in build `streamingProcess: runInteractiveProcess: exec: does not exist (No such file or directory)`
+
+This error could occurs when using `stack` or `cabal` (`Haskell` language building tools) with the `haskell-language-server` if a library or binary is missing in the language server `PATH`.
+
+Depending on how you installed your `Haskell` tools, you will need to add one or more of the following to your Sublime Text `PATH`.
```
Comment (rwols) [inline docs/src/troubleshooting.md:80]: you will need to add one or more of the following to your Sublime Text `PATH`.

-->

you will need to add one or more of the following lines to your login script for bash-like shells on a unix-like operating system.

**Review by rwols — APPROVED** [review]
Yeah, looks good to me. @rchl or @predragnikolic any thoughts?

**Comment by predragnikolic** [comment]
Looking at this :)

**Comment by predragnikolic** [comment]
I'll be a critique on this one, but bare with me :)

I don't thing that the [4 point - Popup error Language server <your_server_language_name> has crashed](https://github.com/sublimelsp/LSP/blob/4308f561621951d04c6ba7d5b86cc6f1eb6fcc82/docs/src/troubleshooting.md#4-popup-error-language-server-your_server_language_name-has-crashed) is necessary, because it was described in the [2 point](https://github.com/sublimelsp/LSP/blob/st4000-exploration/docs/src/troubleshooting.md#2-lsp-cannot-find-my-language-server-no-such-file-or-directory-xyz). Just moving it to a 4 point doesn't make a big difference. :) 

Lets compare [point 3 -  Error in build streamingProcess: runInteractiveProcess: exec: does not exist (No such file or directory)](https://github.com/sublimelsp/LSP/blob/4308f561621951d04c6ba7d5b86cc6f1eb6fcc82/docs/src/troubleshooting.md#3-error-in-build-streamingprocess-runinteractiveprocess-exec-does-not-exist-no-such-file-or-directory) 
with [point 2 - LSP cannot find my language server (No such file or directory: 'xyz')](https://github.com/sublimelsp/LSP/blob/4308f561621951d04c6ba7d5b86cc6f1eb6fcc82/docs/src/troubleshooting.md#2-lsp-cannot-find-my-language-server-no-such-file-or-directory-xyz).
They both try to solve the same issue: `does not exist (No such file or directory)`.

Point 3 is not a generic solution, it is tailored for the Haskell ecosystem, 
while the point 2 has is more generic. 

I am curious to see why the second step didn't help you to solve your problem in the first place? (Was it not clear, was it missing an example, what could we do to make the second step more)

Summary:
I think the [current troubleshooting guide](https://github.com/sublimelsp/LSP/blob/st4000-exploration/docs/src/troubleshooting.md#2-lsp-cannot-find-my-language-server-no-such-file-or-directory-xyz) is simpler, than the one in this PR. 

If you could just pin-point the problems you had and say that a particular paragraph/s wasn't clear in step 2. If that was written a little differently that would have helped me solve my issue.
so to make the current troubleshooting guide more simpler that would be nice :)

**Comment by JeremyBois** [comment]
> I don't thing that the [4 point - Popup error Language server <your_server_language_name> has crashed](https://github.com/sublimelsp/LSP/blob/4308f561621951d04c6ba7d5b86cc6f1eb6fcc82/docs/src/troubleshooting.md#4-popup-error-language-server-your_server_language_name-has-crashed) is necessary, because it was described in the [2 point](https://github.com/sublimelsp/LSP/blob/st4000-exploration/docs/src/troubleshooting.md#2-lsp-cannot-find-my-language-server-no-such-file-or-directory-xyz). Just moving it to a 4 point doesn't make a big difference. :)

Actually I did not get the message `No such file or directory: "xyz"` but only a message indicating the crash of the plugin. So I think a proper section can help a future user. 

Also an alternative solution is added in the document. It consists of updating `env` variable directly in the **LSP** settings for a specific server. This approach allows to only expose a folder for a specific language server.

> Lets compare point 3 - Error in build streamingProcess: runInteractiveProcess: exec: does not exist (No such file or directory) with point 2 - LSP cannot find my language server (No such file or directory: 'xyz'). They both try to solve the same issue: does not exist (No such file or directory).

Again yes, the same solution for another issue. In this case the problem is a missing dependency with a **working** language server. For instance when the binary for the language server and one of its dependency is not in the same folder.

> Point 3 is not a generic solution, it is tailored for the Haskell ecosystem, while the point 2 has is more generic.

Yes I agree with this one. The error message is linked to Haskell in this case but the missing dependency is always specific to the server. Maybe I could remove this part and add a note on missing dependencies issue ?


> If you could just pin-point the problems you had and say that a particular paragraph/s wasn't clear in step 2. The goal is to make the current troubleshooting guide more simpler :)

As said above, I never get `No such file or directory: "xyz"`, so the current troubleshooting was not directly useful.


**Summary:**
  - Point 4 is different from point 2 even if the solution is the same
  - Point 3 is too specific and could be removed
  - A note about a missing language server dependencies is needed
  - Added alternative solution avoid polluting the global user space only for a specific server

**Comment by predragnikolic** [comment]
I would just suggest removing the 3rd point from the troubleshooting guide.
The rest what you said in your summary is ok.

File: docs/src/troubleshooting.md [hunk 3]
```diff
@@ -42,15 +30,60 @@ Adjusting `PATH` can differ based on the operating system and the default shell
...
+
+
+### 3. Popup error `Language server <your_server_language_name> has crashed`
+Assuming that the server is actually installed, and that you can start it from your shell, this issue is likely due to Sublime Text's internal environment not picking up the same `PATH` environment variable as you've configured in your shell.
+
+> **Note** : Language servers may have dependencies that should also be in your `PATH` in addition to the server binary itself.
+
+For instance if you have installed the `haskell-language-server` using [ghcup-hs](https://gitlab.haskell.org/haskell/ghcup-hs) you should expose its specific installation folder `~/.ghcup/bin`. If the build process uses `stack` then it should also be in your `PATH`.
+
+The complete procedure of updating the `PATH` used by Sublime Text depends on your platform and is explained [here](troubleshooting.md#updating-the-path-used-by-lsp-servers).
```
Comment (rchl) [inline docs/src/troubleshooting.md:84]: I don't think this should be here. The reasons for server crashing can be plenty and I don't think a wrong PATH is a frequent reason for those. I'd probably add it in the Haskell section in https://github.com/sublimelsp/LSP/blob/st4000-exploration/docs/src/language_servers.md instead.

File: docs/src/troubleshooting.md [hunk 4]
```diff
@@ -42,15 +30,60 @@ Adjusting `PATH` can differ based on the operating system and the default shell
...
+
+Your client configuration requires two settings to match the document your are editing:
+
+* Scope (eg. `source.php`): Verify this is correct by running "Show Scope Name" from the developer menu.
+* Syntax (eg. `Packages\PHP\PHP.sublime-syntax`): Verify by running `view.settings().get("syntax")` in the console.
+
+### 2. LSP cannot find my language server (`No such file or directory: 'xyz'`)
+Assuming that the server is actually installed, and that you can start it from your shell, this issue is likely due to Sublime Text's internal environment not picking up the same `PATH` environment variable as you've configured in your shell.
+
+The exact changes to make can differ depending on what program you want to expose to Sublime Text. The simplest way is to extend the path like so (replacing `/usr/local/bin` with the path of your choice):
```
Comment (rchl) [inline docs/src/troubleshooting.md:67]: The `The exact changes to make` paragraph (including the code block) should be in the `Updating the PATH used by LSP servers` section. Just maybe add info that this is a Linux specific method.

**Review by rchl — CHANGES_REQUESTED** [review]

**Comment by rwols** [comment]
We can only ask so much from contributors. Maybe we can merge this and then you can make the adjustments, @rchl?

**Review by rwols — APPROVED** [review]

**Comment by rchl** [comment]
> We can only ask so much from contributors. Maybe we can merge this and then you can make the adjustments, @rchl?

I can do that when I find time and courage.

**Pull request merged**